crates/wac-types/src/aggregator.rs

```diff
@@ -165,22 +160,49 @@ impl TypeAggregator {
     ) -> Result<()> {
         let source = &types[id];
         for (name, used) in &source.uses {
+            let source_interface = types[used.interface]
+                .id
+                .as_ref()
+                .context("used type has no interface identifier")?;
+
+            // Validate any existing used type of the same name
             if let Some(existing) = self.types[existing].uses.get(name) {
-                if existing.name != used.name {
-                    bail!("mismatched used type names");
+                let existing_interface = self.types[existing.interface]
+                    .id
+                    .as_ref()
+                    .context("used type has no interface identifier")?;
+
+                // The interface names must match
+                if existing_interface != source_interface {
+                    bail!("cannot merge used type `{name}` as it is expected to be from interface `{existing_interface}` but it is from interface `{source_interface}`");
                 }
 
-                self.merge_interface(existing.interface, types, used.interface, checker)?;
-                continue;
+                // The types must be exported with the same name
+                if existing.name != used.name {
+                    bail!("cannot merge used type `{name}` as the export names are mismatched");
+                }
             }
 
-            let remapped = UsedType {
-                interface: self.remap_interface(types, used.interface),
-                name: used.name.clone(),
-            };
-
-            let prev = self.types[existing].uses.insert(name.clone(), remapped);
-            assert!(prev.is_none());
+            // Remap the used interface; this will handle merging if we've seen the interface before
+            let remapped = self.remap_interface(types, used.interface, checker)?;
+            match self.types[existing].uses.get(name) {
+                Some(existing) => {
+                    assert_eq!(
+                        existing.interface, remapped,
+                        "expected a merge to have occurred"
+                    );
+                }
+                None => {
+                    let prev = self.types[existing].uses.insert(
+                        name.clone(),
+                        UsedType {
+                            interface: remapped,
+                            name: used.name.clone(),
+                        },
+                    );
+                    assert!(prev.is_none());
```

Comment:
This `assert` seems unnecessary since we're still within the check that that type doesn't exist in `self.types`.